proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1189 @@
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+  {-# LANGUAGE TemplateHaskell #-}
+  module Main where
+
+  import splice A (foo)  -- foo :: Int -> Q Exp
+  import B (bar)         -- bar :: Int -> Q Exp
+
+  x = $(foo 25) -- Accepted
+  y = $(bar 33) -- Error: bar imported at level 0 but used at level -1
+
+Thus:
+
+1. When compiling module ``Main``, even though ``TemplateHaskell`` is enabled,
+   only identifiers from module ``A`` will be used in top-level splices so
+   only ``A`` (and its dependencies) needs to compiled to object code before starting to compile ``Main``.
+2. When cross-compiling, ``A`` needs to be built only for the host and ``B``
+   only for the target.
+
+If the same module is needed to be used at different levels
+then two import declarations can be used::
+
+  import C
+  import splice C
+
+Splice imports example: ``printf``
+##################################
+
+Let ``printf :: String -> Q Exp`` be defined in ``Printf``, such that the
+arguments received by ``printf`` applied to a formatting string is determined at
+compile time based on the format specifiers within the string::
+
+    $(printf "Error: %s on line %d") "test" 123 :: String
+
+The following program is rejected::
+
+    {-# LANGUAGE ExplicitLevelImports #-}
+
+    import Printf (printf)
+
+    -- Error: printf imported at level 0 but used at level -1
+    x = $(printf "Error: %s on line %d") "test" 123 :: String
+
+because ``printf`` was imported "normally" at the default level 0 and thus
+cannot occur within a top-level splice (at level -1). For this program to be
+level-correct, ``printf`` must be imported at level -1 to be used within a
+top-level splice::
+
+    {-# LANGUAGE ExplicitLevelImports #-}
+
+    import splice Printf (printf)
+
+    -- accepted!
+    x = $(printf "Error: %s on line %d") "test" 123 :: String
+
+Splice-importing ``Printf`` makes it clear to both humans and compilers that
+``printf`` will only be required at compile time, since it will only be used within top-level splices.
+
+
+
+Quote imports
+#############
+
+A "quote" import is prefixed with ``quote``.  In this example, identifiers from
+``A`` can be used **only** in quotes, while identifiers from ``A`` **cannot** be
+used in quotes or splices::
+
+  {-# LANGUAGE ExplicitLevelImports #-}
+  {-# LANGUAGE TemplateHaskell #-}
+  module Main where
+
+  import quote A (foo)  -- foo :: Int -> Int
+  import B (bar)        -- bar :: Int -> Int
+
+  x = [| foo 25 |] -- Accepted
+  y = [| bar 33 |] -- Error: bar imported at level 0 but used at level 1
+
+When a quote such as ``x = [| foo 25 |]`` is spliced, i.e. ``z = $(x)``,
+its contents will be needed to execute the program at runtime (``z = foo 25``,
+so evaluating ``z`` at runtime requires ``foo`` to be available).
+
+
+
+Module Stages
+#############
+
+Modules are compiled at a specific stage. Levels within a module are interpreted
+as offsets to the specific stage at which the module is being compiled.
```

Comment:
You should strengthen the wording here, specifying that compiling at stages is an application of the proposal, rather than part of the proposal itself. Also,  are stages a concept which already exists in Cabal today, or is it something that we'll want in the future?

Reply:
They are something that we'll want in the future

Qualified goals however are a bit siilar.

Reply:
I have updated the proposal with some clarification around this.

GHC has a basic notion of a stage, but it is left quite implicit curently. Cabal doesn't know anything about stages.